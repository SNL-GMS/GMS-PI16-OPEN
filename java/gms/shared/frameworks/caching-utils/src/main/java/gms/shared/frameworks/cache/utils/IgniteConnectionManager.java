package gms.shared.frameworks.cache.utils;

import com.google.common.base.Preconditions;
import gms.shared.frameworks.systemconfig.SystemConfig;
import gms.shared.utilities.logging.StructuredLoggingWrapper;
import org.apache.ignite.Ignite;
import org.apache.ignite.IgniteCache;
import org.apache.ignite.Ignition;
import org.apache.ignite.configuration.CacheConfiguration;
import org.apache.ignite.configuration.IgniteConfiguration;
import org.apache.ignite.events.EventType;
import org.apache.ignite.logger.log4j.Log4JLogger;
import org.apache.ignite.spi.eventstorage.memory.MemoryEventStorageSpi;
import org.slf4j.LoggerFactory;

import javax.cache.Cache;
import java.io.Closeable;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * The {@link IgniteConnectionManager} is used to create {@link Cache} instances from an Ignite node. This way a client just
 * need to adhere to accessing the cache via the JCache API rather than using an Ignite-specific query mechanism.
 * <p>
 * The class implements the {@link Closeable} interface in order to allow us to close the Ignite node upon shutdown of a
 * given application or use try-with-resources when scoping down the access to an Ignite node specifically.
 */
public class IgniteConnectionManager {

  private static final String NAMESPACE_LOCATION = "/var/run/secrets/kubernetes.io/serviceaccount/namespace";
  private static final String CONFIG_IGNITE_INSTANCE_NAME = "ignite-instance-name";
  private static Ignite ignite;
  private static final StructuredLoggingWrapper logger = StructuredLoggingWrapper
      .create(LoggerFactory.getLogger(IgniteConnectionManager.class));
  private IgniteConnectionManager() {
  } //private constructor for sonarlint

  /**
   * Create {@link IgniteConnectionManager} that controls configuration of cache and registers creator as a server node
   * server nodes participate in the cluster to store data
   * <p>
   * Clients can simply call IgniteConnectionManager.getInstance().getOrCreateCache(CacheName.{name}); to retrieve data from
   * the cluster
   *
   * @param systemConfig - system config to setup cache cluster
   * @param cacheList - caches to create in the cluster.  The node calling this will host (some of) the data
   */
  public static synchronized void create(SystemConfig systemConfig, List<CacheInfo> cacheList) {
    logger.info("Initializing ignite cache");
    Objects.requireNonNull(cacheList);
      Objects.requireNonNull(systemConfig);
    Preconditions.checkState(ignite == null,
        "IgniteConnectionManager.create() should only be called one time, during application startup");
    createConfiguration(systemConfig, cacheList);
    Runtime.getRuntime().addShutdownHook(new Thread(() -> {
      if (ignite != null) {
        logger.info("Closing Ignite connections");
        close();
      }
    }));
  }

  private static IgniteConfiguration createConfiguration(SystemConfig systemConfig, List<CacheInfo> cacheList) {
    Map<String, Object> nodeAttributes = new HashMap<>();
    cacheList.forEach(cache -> nodeAttributes.put(cache.getNodeAttr(), true));
    var igniteLogger = new Log4JLogger();
    MemoryEventStorageSpi eventStorageSpi = new MemoryEventStorageSpi()
      .setExpireCount(60000);

    String igniteInstanceName = systemConfig.getValue(CONFIG_IGNITE_INSTANCE_NAME);

    IgniteConfiguration cfg = new IgniteConfiguration()
      .setIgniteInstanceName(igniteInstanceName)
      .setUserAttributes(nodeAttributes)
      .setGridLogger(igniteLogger)
      .setIncludeEventTypes(EventType.EVT_CACHE_OBJECT_PUT, EventType.EVT_CACHE_OBJECT_READ,
        EventType.EVT_CACHE_OBJECT_REMOVED, EventType.EVT_NODE_JOINED, EventType.EVT_NODE_LEFT)
      .setEventStorageSpi(eventStorageSpi);
    if (cacheList.isEmpty()) {
      cfg.setClientMode(true);
    }
    cfg.setCacheConfiguration(initializeCaches(cacheList).toArray(CacheConfiguration[]::new));
    ignite = Ignition.start(cfg);

    return cfg;
  }

  private static <T> Collection<CacheConfiguration> initializeCaches(List<CacheInfo> cacheList) {
    return cacheList.stream()
      .map(IgniteConnectionManager::createCacheFromCacheInfo)
      .collect(Collectors.toList());

  }

  private static <T, U> CacheConfiguration<T, U> createCacheFromCacheInfo(CacheInfo cache) {
    logger.info("Getting or creating cache: " + cache.getCacheName());
    CacheConfiguration<T, U> cacheCfg = new CacheConfiguration<>();
    cacheCfg.setName(cache.getCacheName());
    cacheCfg.setCacheMode(cache.getCacheMode());
    cacheCfg.setAtomicityMode(cache.getCacheAtomicityMode());
    cacheCfg.setOnheapCacheEnabled(cache.isOnHeap());
    cacheCfg.setNodeFilter(cache.getNodeFilter());
    if (cache.getEvictionPolicy().isPresent()) {
      cacheCfg.setEvictionPolicyFactory(cache.getEvictionPolicy().get());
    }
    return cacheCfg;
    //return ignite.getOrCreateCache(cacheCfg);
  }

  /**
   * retrieve a given {@link Cache} by the cacheName.
   *
   * @param cache name of the {@link Cache} instance to retrieve.
   * @param <K> the type of the key to retrieve a {@link Cache.Entry} instance.
   * @param <V> the type of the value of the retrieved {@link Cache.Entry} instance.
   * @return instance of {@link Cache} representing the requested Cache retrieved.  null if all server nodes are down
   */
  public static <K, V> IgniteCache<K, V> getOrCreateCache(CacheInfo cache) {
    Objects.requireNonNull(cache);
    Objects.requireNonNull(ignite, "Ignite has not been initialized.  Did you call create()?");
    return ignite.getOrCreateCache(createCacheFromCacheInfo(cache));
  }

  public static void close() {
    if(ignite != null)
      ignite.close();
    IgniteConnectionManager.ignite = null;
  }
  private static String getKubernetesNamespace() {
    var nameSpaceFile = Path.of(NAMESPACE_LOCATION);
    var namespace =  "default";
    try {
      namespace = Files.readString(nameSpaceFile);
    } catch (IOException e) {
      logger.error("could not load namespace from " + NAMESPACE_LOCATION, e);
    }
    logger.info("Ignite kubernetesIpFinder using namespace " + namespace);
    return namespace;
  }
}

