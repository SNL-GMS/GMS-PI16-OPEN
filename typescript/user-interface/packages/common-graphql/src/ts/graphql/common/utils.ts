/**
 * Helper function to remove the __typename from object.
 * Gateway GraphQL doesn't like it.
 *
 * @param object Generic object to strip __typename out of
 * @returns new object copied from param with __typename stripped out of
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/no-explicit-any
export function removeTypeName(object: any): any {
  const newObj = {};
  // eslint-disable-next-line no-restricted-syntax
  for (const key in object) {
    if (object[key] instanceof Object) {
      newObj[key] = removeTypeName(object[key]);
    } else if (key !== '__typename') {
      newObj[key] = object[key];
    }
  }
  return newObj;
}
