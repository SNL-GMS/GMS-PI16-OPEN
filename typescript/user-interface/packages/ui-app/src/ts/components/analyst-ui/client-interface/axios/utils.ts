import { CommonTypes, WorkflowTypes } from '@gms/common-model';
import { MILLISECONDS_IN_SECOND, toEpochSeconds } from '@gms/common-util';
import Immutable from 'immutable';
import moment from 'moment';

import { queryCache } from '~components/client-interface';

import { getWorkflowIntervalQueryConfig } from './queries/workflow-interval-query';

// TODO once our query key logic is cleaned up, will not need stageNames and timeRange
export const cleanUpWorkflow = (
  latestIntervalEndTime: number,
  operationalTimeDuration: number,
  stageNames: string[],
  timeRange: CommonTypes.TimeRange
): void => {
  const startTime = toEpochSeconds(
    moment((latestIntervalEndTime - operationalTimeDuration) * MILLISECONDS_IN_SECOND)
      .utc()
      .startOf('hour')
      .toISOString()
  );
  const { queryKey } = getWorkflowIntervalQueryConfig(stageNames, timeRange);
  let cachedWorkflowIntervalResultMap: Immutable.Map<
    string,
    WorkflowTypes.StageInterval[]
  > = queryCache.getQueryData(queryKey);
  // TODO maybe optimize with bailing once finding a valid endTime
  let foundStaleEntries = false;
  cachedWorkflowIntervalResultMap.forEach((entries, key) => {
    const newEntries = entries
      .map(entry => {
        if (entry.startTime < startTime) {
          foundStaleEntries = true;
          return undefined;
        }
        return entry;
      })
      .filter(entry => entry !== undefined);
    cachedWorkflowIntervalResultMap = cachedWorkflowIntervalResultMap.set(key, newEntries);
  });
  if (foundStaleEntries) {
    queryCache.setQueryData(queryKey, cachedWorkflowIntervalResultMap);
  }
};
