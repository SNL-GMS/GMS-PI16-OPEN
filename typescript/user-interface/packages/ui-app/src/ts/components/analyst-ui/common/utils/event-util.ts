import { EventTypes, SignalDetectionTypes } from '@gms/common-model';
import flatMap from 'lodash/flatMap';

import { userPreferences } from '~analyst-ui/config/user-preferences';

/**
 * Get the current open event
 *
 * @param openEventId Id of open event
 * @param eventsInTimeRangeQuery eventsInTimeRangeQuery
 */
export function getOpenEvent(
  openEventId: string | undefined,
  events: EventTypes.Event[]
): EventTypes.Event | undefined {
  const event =
    openEventId && events && events.length > 0 ? events.find(e => e.id === openEventId) : undefined;
  return event && event.currentEventHypothesis && event.currentEventHypothesis.eventHypothesis
    ? event
    : undefined;
}

/**
 * Gets the signal detections associated to the given event
 *
 * @param event the open even to get sd's for
 * @param signalDetections signalDetections to look through
 */
export function getAssocSds(
  event: EventTypes.Event,
  signalDetections: SignalDetectionTypes.SignalDetection[]
): SignalDetectionTypes.SignalDetection[] {
  return event && event.currentEventHypothesis && event.currentEventHypothesis.eventHypothesis
    ? flatMap(event.currentEventHypothesis.eventHypothesis.signalDetectionAssociations, assocSD => {
        const maybeSD = signalDetections.find(
          sd =>
            assocSD.signalDetectionHypothesis.id === sd.currentHypothesis.id && !assocSD.rejected
        );
        if (maybeSD) {
          return maybeSD;
        }
        return undefined;
      }).filter(assocSD => assocSD !== undefined)
    : [];
}

/**
 * Returns the latest location solution set for the provided event.
 *
 * @param event an event
 */
export function getLatestLocationSolutionSet(
  event: EventTypes.Event
): EventTypes.LocationSolutionSet | undefined {
  return event && event.currentEventHypothesis && event.currentEventHypothesis.eventHypothesis
    ? // eslint-disable-next-line @typescript-eslint/no-use-before-define
      getLatestLocationSolutionSetForEventHypothesis(event.currentEventHypothesis.eventHypothesis)
    : undefined;
}

/**
 * Returns the latest location solution set for the provided event hypothesis.
 *
 * @param eventHypothesis an event hypothesis
 */
export function getLatestLocationSolutionSetForEventHypothesis(
  eventHypothesis: EventTypes.EventHypothesis
): EventTypes.LocationSolutionSet {
  return eventHypothesis &&
    eventHypothesis.locationSolutionSets &&
    eventHypothesis.locationSolutionSets.length > 0
    ? eventHypothesis.locationSolutionSets.reduce((prev, curr) => {
        if (curr.count > prev.count) {
          return curr;
        }
        return prev;
      }, eventHypothesis.locationSolutionSets[0])
    : undefined;
}

/**
 * Gets the default preferred location id for an event based off the config
 * Or if not found returns the id for the preferred location solution
 *
 * @param locationSolutionSet Location Solution Set to  get default preferred location solution from
 */
export function getPreferredDefaultLocationId(
  locationSolutionSet: EventTypes.LocationSolutionSet
): string | undefined {
  if (!locationSolutionSet.locationSolutions || locationSolutionSet.locationSolutions.length < 1) {
    return undefined;
  }

  let toReturn: string;
  // A for loop is used so we can break
  // eslint-disable-next-line @typescript-eslint/prefer-for-of
  for (
    let i = 0;
    i < userPreferences.location.preferredLocationSolutionRestraintOrder.length;
    // eslint-disable-next-line no-plusplus
    i++
  ) {
    const dr = userPreferences.location.preferredLocationSolutionRestraintOrder[i];
    const maybeLS = locationSolutionSet.locationSolutions.find(
      ls => ls.locationRestraint.depthRestraintType === dr
    );
    if (maybeLS) {
      toReturn = maybeLS.id;
      break;
    }
  }
  if (toReturn) {
    return toReturn;
  }
  return locationSolutionSet[0].id;
}

/**
 * Gets the default preferred location id for an event based off the config
 * Or if not found returns the id for the preferred location solution
 *
 * @param eventHypothesis Event hypothesis to get default preferred location solution from
 */
export function getPreferredLocationSolutionIdFromEventHypothesis(
  eventHypothesis: EventTypes.EventHypothesis
): string {
  if (!eventHypothesis.locationSolutionSets || eventHypothesis.locationSolutionSets.length < 1) {
    return undefined;
  }
  const set = getLatestLocationSolutionSetForEventHypothesis(eventHypothesis);
  return getPreferredDefaultLocationId(set);
}

/**
 * Gets the distance to all stations for the given location solution id
 *
 * @param event event with location solution
 * @param locationSolutionId location solution id
 */
export function getDistanceToStationsForLocationSolutionId(
  event: EventTypes.Event,
  locationSolutionId: string
): EventTypes.LocationToStationDistance[] {
  if (!event) {
    return [];
  }
  const locationSolutions = flatMap(
    event.currentEventHypothesis.eventHypothesis.locationSolutionSets,
    lss => lss.locationSolutions
  );
  const maybeLs = locationSolutions.find(ls => ls.id === locationSolutionId);
  if (maybeLs) {
    return maybeLs.locationToStationDistances;
  }
  return event.currentEventHypothesis.eventHypothesis.preferredLocationSolution.locationSolution
    .locationToStationDistances;
}

/**
 * Checks if event hypothesis has changed or if the number of location solution sets has changed
 *
 * @param event current props event
 * @param prevEvent previous props event
 *
 * @returns boolean if changes have occurred
 */
export function shouldUpdateSelectedLocationSolution(
  prevEvent: EventTypes.Event,
  event: EventTypes.Event
): boolean {
  return (
    prevEvent.currentEventHypothesis.eventHypothesis.id !==
      event.currentEventHypothesis.eventHypothesis.id ||
    prevEvent.currentEventHypothesis.eventHypothesis.locationSolutionSets.length !==
      event.currentEventHypothesis.eventHypothesis.locationSolutionSets.length
  );
}
