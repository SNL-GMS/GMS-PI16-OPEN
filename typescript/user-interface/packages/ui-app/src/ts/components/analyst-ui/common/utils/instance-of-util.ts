import {
  ChannelSegmentTypes,
  FkTypes,
  SignalDetectionTypes,
  WaveformTypes
} from '@gms/common-model';

/**
 * Checks if FK spectra channel segment
 *
 * @param object Channel Segment
 * @returns boolean
 */
// eslint-disable-next-line max-len
export function isFkSpectraChannelSegment(
  object: ChannelSegmentTypes.ChannelSegment<ChannelSegmentTypes.TimeSeries>
): object is ChannelSegmentTypes.ChannelSegment<FkTypes.FkPowerSpectra> {
  return object.timeseriesType === ChannelSegmentTypes.TimeSeriesType.FK_SPECTRA;
}

/**
 * Checks if Signal detection InstantMeasurementValue
 *
 * @param object FeatureMeasurementValue
 * @returns boolean
 */
export function isSdInstantMeasurementValue(
  object: any
): object is SignalDetectionTypes.InstantMeasurementValue {
  return object.value !== undefined && object.standardDeviation !== undefined;
}

/**
 * Checks if Signal detection AmplitudeMeasurementValue
 *
 * @param object FeatureMeasurementValue
 * @returns boolean
 */
export function isSdFeatureMeasurementValue(
  object: any
): object is SignalDetectionTypes.AmplitudeMeasurementValue {
  return (
    object.amplitude !== undefined && object.period !== undefined && object.startTime !== undefined
  );
}

/**
 * Creates/Returns an unfiltered waveform filter
 */
export function createUnfilteredWaveformFilter(): WaveformTypes.WaveformFilter {
  return (WaveformTypes.UNFILTERED_FILTER as any) as WaveformTypes.WaveformFilter;
}
