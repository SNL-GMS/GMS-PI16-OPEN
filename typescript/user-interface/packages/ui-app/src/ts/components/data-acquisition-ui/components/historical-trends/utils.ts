import { SohTypes } from '@gms/common-model';
import { ChartTypes } from '@gms/ui-core-components';
import { DistinctColorPalette, hslToHex } from '@gms/ui-util';

import { dataAcquisitionUIConfig } from '~components/data-acquisition-ui/config';
import { BarLineChartData } from '~components/data-acquisition-ui/shared/chart/types';

const THOUSAND = 1000;

/** Returns based on the type provided */
export const getName = (type: SohTypes.SohMonitorType): string => {
  const lowercase: string = type.valueOf().toLowerCase();
  return lowercase.charAt(0).toUpperCase() + lowercase.slice(1);
};

/** Returns the label based on the type provided */
export const getLabel = (type: SohTypes.SohMonitorType): string => `${getName(type)} data`;

/** Returns the data type of the data based on the type provided */
export const getDataType = (type: SohTypes.SohMonitorType): string =>
  // !update this function when adding in new types
  type === SohTypes.SohMonitorType.MISSING ? '%' : 's';

/** A helper function for formatting chart tic labels; formats numbers to represent thousandths */
export const chartTickFormatForThousands = (value: number): string | number =>
  value >= THOUSAND ? `${value / THOUSAND}k` : value;

/** Returns the bar chart Y axis label based on the type provided */
export const barChartYAxisLabel = (type: SohTypes.SohMonitorType): string =>
  `Average ${getName(type)} (${getDataType(type)})`;

/** Returns the bar chart X axis label */
export const barChartXAxisLabel = (): string => 'Channel Name';

/** The bar chart Y axis tic formatter */
export const barChartYAxisTicFormat = chartTickFormatForThousands;

/** The bar chart X axis formatter */
export const barChartXAxisTicFormat = (stationName: string) => (name: string): string =>
  name && name.replace ? name.replace(`${stationName}.`, '') : name;

/** Returns the line chart Y axis label based on the type provided */
export const lineChartYAxisLabel = (type: SohTypes.SohMonitorType): string =>
  `${getName(type)} (${getDataType(type)})`;

/** Returns the line chart X axis label */
export const lineChartXAxisLabel = (): string => 'Time';

export const getChartHeight = (targetHeightPx: number): number =>
  Math.max(targetHeightPx, dataAcquisitionUIConfig.dataAcquisitionUserPreferences.minChartHeightPx);

/**
 * Returns the chart data needed to populate the bar and line charts for
 * the historical data.
 *
 * @param props historical trend history display props
 * @param context the historical soh query data
 */
export const getChartData = (
  monitorType: SohTypes.SohMonitorType,
  station: SohTypes.UiStationSoh,
  uiHistoricalSoh: SohTypes.UiHistoricalSohAsTypedArray,
  colorPalette: DistinctColorPalette
): BarLineChartData => {
  const monitorValues = uiHistoricalSoh?.monitorValues;

  const lineDefs: ChartTypes.WeavessLineDefinition[] = [];
  const barDefs: ChartTypes.BarDefinition[] = [];

  let categories: { x: string[]; y: string[] } = { x: [], y: [] };

  // Iterate through each channel monitor type building the line definition
  // and bar definition and add them to their respective arrays
  if (monitorValues) {
    categories = {
      x: monitorValues.map(mV => mV.channelName),
      y: []
    };
    // Iterate through each channel monitor type building the
    // bar definition and add them to the arrays
    monitorValues.forEach(mV => {
      // Add the line definition for this channel
      lineDefs.push({
        id: mV.channelName,
        color: hslToHex(colorPalette.getColor(mV.channelName)),
        values: mV.values,
        average: mV.average
      });
      // Add the bar definition for this channel
      barDefs.push({
        id: mV.channelName,
        color: colorPalette.getColorString(mV.channelName),

        value: {
          x: categories.x.find(name => name === mV.channelName),
          y: mV.average
        }
      });
    });
  }

  // Valid the bar definitions
  barDefs.forEach(
    (barDef, index) =>
      barDef.id !== categories.x[index] &&
      // eslint-disable-next-line no-console
      console.error('channel names of historical data may not match actual values')
  );

  const statuses = station?.channelSohs?.map(c =>
    c.allSohMonitorValueAndStatuses.find(v => v.monitorType === monitorType)
  );

  const thresholdsMarginal = statuses?.filter(s => s !== undefined).map(s => s.thresholdMarginal);
  const thresholdsBad = statuses?.filter(s => s !== undefined).map(s => s.thresholdBad);
  return {
    categories,
    lineDefs,
    barDefs,
    thresholdsMarginal,
    thresholdsBad,
    minAndMax: uiHistoricalSoh?.minAndMax
  };
};
