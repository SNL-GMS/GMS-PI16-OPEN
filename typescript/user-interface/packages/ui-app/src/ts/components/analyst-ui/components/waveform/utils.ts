import { CommonTypes, EventTypes } from '@gms/common-model';
import { WeavessTypes } from '@gms/weavess-core';

import { isSdInstantMeasurementValue } from '~analyst-ui/common/utils/instance-of-util';

export interface Offset {
  channelId: string;
  offset: number;
}

/**
 * Calculate offsets based on station with earliest arrival.
 */
export const calculateOffsets = (
  featurePredictions: EventTypes.FeaturePrediction[],
  baseChannelId: string,
  phaseToOffset: string
): Offset[] => {
  if (!featurePredictions || featurePredictions.length <= 0) {
    return [];
  }
  const filteredPredictions: EventTypes.FeaturePrediction[] = featurePredictions.filter(
    fp => fp.phase === phaseToOffset
  );

  // If no FPs were found for the phase selected, return nothing
  if (filteredPredictions.length <= 0) {
    return [];
  }

  filteredPredictions.sort((a, b) => {
    if (
      isSdInstantMeasurementValue(a.predictedValue) &&
      isSdInstantMeasurementValue(b.predictedValue)
    ) {
      const aValue = a.predictedValue.value;
      const bValue = b.predictedValue.value;
      return aValue - bValue;
    }
    return 0;
  });

  const offsets: Offset[] = [];
  const baseChannelFeaturePrediction = filteredPredictions.find(
    fp => fp.channelName === baseChannelId
  );

  if (
    baseChannelFeaturePrediction &&
    isSdInstantMeasurementValue(baseChannelFeaturePrediction.predictedValue)
  ) {
    const baseTime: number = baseChannelFeaturePrediction.predictedValue.value;

    filteredPredictions.forEach(sortedFeaturePrediction => {
      if (isSdInstantMeasurementValue(sortedFeaturePrediction.predictedValue)) {
        const offset: Offset = {
          channelId: sortedFeaturePrediction.channelName,
          offset: baseTime - sortedFeaturePrediction.predictedValue.value
        };
        offsets.push(offset);
      }
    });
  }
  return offsets;
};

/**
 * TODO: Remove if/when we convert the UI TimeRange to use the same property keys.
 * Converts a UI time range to the Weavess format.
 *
 * @param timeRange a time range in the common model format
 * @returns a timeRange in the weavess format
 */
export const convertToWeavessTimeRange = (
  timeRange: CommonTypes.TimeRange
): WeavessTypes.TimeRange => ({
  startTimeSecs: timeRange.startTimeSecs,
  endTimeSecs: timeRange.endTimeSecs
});
