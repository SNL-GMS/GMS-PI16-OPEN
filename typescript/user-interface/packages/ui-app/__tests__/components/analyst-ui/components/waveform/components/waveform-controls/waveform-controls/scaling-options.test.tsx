/* eslint-disable @typescript-eslint/no-empty-function */
import Enzyme from 'enzyme';
import * as React from 'react';

import {
  AmplitudeScalingOptions,
  useScalingOptions
} from '../../../../../../../../src/ts/components/analyst-ui/components/waveform/components/waveform-controls/scaling-options';
import { waitForComponentToPaint } from '../../../../../../../utils/general-utils';

jest.mock(
  '~components/client-interface/axios/queries/processing-analyst-configuration-query',
  () => ({
    useProcessingAnalystConfigurationQuery: jest.fn(() => ({
      data: { fixedAmplitudeScaleValues: [1, 2] }
    }))
  })
);

describe('Scaling Options Toolbar Item', () => {
  it('should exist', () => {
    expect(useScalingOptions).toBeDefined();
  });
  it('should return a ToolbarItem element that matches a snapshot', async () => {
    const ScalingOptionWrapper: React.FC = () => {
      const scalingOptionToolbarItem = useScalingOptions(
        0,
        AmplitudeScalingOptions.AUTO,
        1,
        () => {},
        () => {}
      );
      return scalingOptionToolbarItem.toolbarItem.element;
    };
    const wrapper = Enzyme.mount(<ScalingOptionWrapper />);
    await waitForComponentToPaint(wrapper);
    expect(wrapper).toMatchSnapshot();
  });
});
