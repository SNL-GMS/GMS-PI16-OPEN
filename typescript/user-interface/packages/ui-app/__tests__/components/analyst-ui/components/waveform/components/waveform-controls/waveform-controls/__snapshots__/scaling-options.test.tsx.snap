// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scaling Options Toolbar Item should return a ToolbarItem element that matches a snapshot 1`] = `
<ScalingOptionWrapper>
  <Blueprint3.Popover
    boundary="scrollParent"
    captureDismiss={false}
    content={
      <ScalingOptionSelector
        ampScaleOption="Auto"
        fixedScaleOptions={
          Array [
            "Current",
            1,
            2,
          ]
        }
        fixedScaleVal={1}
        setAmpScaleOption={[Function]}
        setFixedScaleVal={[Function]}
      />
    }
    defaultIsOpen={false}
    disabled={false}
    fill={false}
    hasBackdrop={false}
    hoverCloseDelay={300}
    hoverOpenDelay={150}
    inheritDarkTheme={true}
    interactionKind="click"
    minimal={false}
    modifiers={Object {}}
    openOnTargetFocus={true}
    position="auto"
    targetTagName="span"
    transitionDuration={300}
    usePortal={true}
    wrapperTagName="span"
  >
    <Manager>
      <span
        className="bp3-popover-wrapper"
      >
        <Reference
          innerRef={[Function]}
        >
          <InnerReference
            innerRef={[Function]}
            setReferenceNode={[Function]}
          >
            <Blueprint3.ResizeSensor
              onResize={[Function]}
            >
              <span
                className="bp3-popover-target"
                onClick={[Function]}
              >
                <Blueprint3.Button
                  alignText="left"
                  className=""
                  key=".0"
                  rightIcon="caret-down"
                  style={
                    Object {
                      "width": 72,
                    }
                  }
                  title="Waveform amplitude scaling options"
                  value="Auto"
                >
                  <button
                    className="bp3-button bp3-align-left"
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    style={
                      Object {
                        "width": 72,
                      }
                    }
                    title="Waveform amplitude scaling options"
                    type="button"
                    value="Auto"
                  >
                    <Blueprint3.Icon
                      key="leftIcon"
                    />
                    <span
                      className="bp3-button-text"
                      key="text"
                    >
                      Auto
                    </span>
                    <Blueprint3.Icon
                      icon="caret-down"
                      key="rightIcon"
                    >
                      <span
                        className="bp3-icon bp3-icon-caret-down"
                        icon="caret-down"
                      >
                        <svg
                          data-icon="caret-down"
                          height={16}
                          viewBox="0 0 16 16"
                          width={16}
                        >
                          <desc>
                            caret-down
                          </desc>
                          <path
                            d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
                            fillRule="evenodd"
                            key="0"
                          />
                        </svg>
                      </span>
                    </Blueprint3.Icon>
                  </button>
                </Blueprint3.Button>
              </span>
            </Blueprint3.ResizeSensor>
          </InnerReference>
        </Reference>
        <Blueprint3.Overlay
          autoFocus={true}
          backdropClassName="bp3-popover-backdrop"
          backdropProps={Object {}}
          canEscapeKeyClose={true}
          canOutsideClickClose={true}
          enforceFocus={true}
          hasBackdrop={false}
          isOpen={false}
          lazy={true}
          onClose={[Function]}
          transitionDuration={300}
          transitionName="bp3-popover"
          usePortal={true}
        />
      </span>
    </Manager>
  </Blueprint3.Popover>
</ScalingOptionWrapper>
`;
